sapio-psbt/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+#[serde(untagged)]
+pub enum PSBTApi {
+    Finished {
+        completed: bool,
+        hex: String,
+    },
+    NotFinished {
+        completed: bool,
+        psbt: String,
+        error: String,
+        errors: Vec<String>,
+    },
+}
+
+pub fn finalize_psbt_format_api(psbt: PartiallySignedTransaction) -> PSBTApi {
+    let secp = Secp256k1::new();
+    psbt.finalize(&secp)
+        .map(|tx| {
+            let hex = bitcoin::consensus::encode::serialize_hex(&tx.extract_tx());
+            PSBTApi::Finished {
+                completed: true,
+                hex,
+            }
+        })
+        .unwrap_or_else(|(psbt, errors)| {
+            let errors: Vec<_> = errors.iter().map(|e| format!("{:?}", e)).collect();
+            let encoded_psbt = base64::encode(serialize(&psbt));
+            PSBTApi::NotFinished {
+                completed: false,
+                psbt: encoded_psbt,
+                error: "Could not fully finalize psbt".into(),
+                errors,
+            }
+        })
+}
+
+pub async fn read_key_from_file(
+    file_name: &std::ffi::OsStr,
+) -> Result<ExtendedPrivKey, Box<dyn Error>> {
+    let buf = tokio::fs::read(file_name).await?;
+    Ok(ExtendedPrivKey::decode(&buf)?)
+}
+
+pub fn sign(
+    xpriv: ExtendedPrivKey,
+    mut psbt: PartiallySignedTransaction,
+    hash_ty: bitcoin::SchnorrSighashType,
+) -> Result<Vec<u8>, PSBTSigningError> {
+    sign_psbt_inplace(&xpriv, &mut psbt, &Secp256k1::new(), hash_ty)?;
+    let bytes = serialize(&psbt);
+    Ok(bytes)
+}
+pub async fn show_pubkey(input: &std::ffi::OsStr) -> Result<(), Box<dyn Error>> {
+    let buf = tokio::fs::read(input).await?;
+    let xpriv = ExtendedPrivKey::decode(&buf)?;
+    println!("{}", ExtendedPubKey::from_priv(&Secp256k1::new(), &xpriv));
+    Ok(())
+}
+pub fn new_key(network: &str, out: &std::ffi::OsStr) -> Result<(), Box<dyn Error>> {
+    let entropy: [u8; 32] = rand::thread_rng().gen();
+    let xpriv = ExtendedPrivKey::new_master(Network::from_str(network)?, &entropy)?;
+    std::fs::write(out, &xpriv.encode())?;
+    println!("{}", ExtendedPubKey::from_priv(&Secp256k1::new(), &xpriv));
+    Ok(())
+}
+
+pub fn sign_psbt_inplace(
+    xpriv: &ExtendedPrivKey,
+    psbt: &mut PartiallySignedTransaction,
+    secp: &Secp256k1<All>,
+    hash_ty: bitcoin::SchnorrSighashType,
+) -> Result<(), PSBTSigningError> {
+    sign_psbt_input_inplace(xpriv, psbt, secp, 0, hash_ty)
+}
+
+#[derive(Debug, Clone)]
+pub enum PSBTSigningError {
+    NoUTXOAtIndex(usize),
+    NoInputAtIndex(usize),
+}
+
+impl Display for PSBTSigningError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+impl Error for PSBTSigningError {}
+
+pub fn sign_psbt_input_inplace(
+    xpriv: &ExtendedPrivKey,
+    psbt: &mut PartiallySignedTransaction,
+    secp: &Secp256k1<All>,
+    idx: usize,
+    hash_ty: bitcoin::SchnorrSighashType,
+) -> Result<(), PSBTSigningError> {
```

Comment:
If you take an owned PSBT and return the new one you will improve the API I think. Allowing you to dotchain these together. I generally prefer owned arguments that return the structure again if it doesn't make the API unwieldy and in this case it seems it would improve because the success part of the result is unit.

Reply:
I think that should be saved for a PsbtExt trait that offers these as an interface v.s. the raw APIs?

Although I did have the owned interface earlier, and refactored to this form. The main "oddity" with the owned API is that you need to be able to retry things, which means your API needs to return the psbt again, i.e. Result<PSBT, (Error, PSBT)> so that you can continue trying things with the PSBT that might fail. But then it's not clear what state the (Error, PSBT) PSBT is in... has it been mutated at all? so then internally either the function has to preallocate a copy, or you need to somehow undo the changes...

feels simpler to leave it up to the caller to decide?

Reply:
IMO there are only two sensible things to do for failure.

1. The original structure is still in a valid state, meaning you can return it in the error side of the Result
2. The original structure is *not* in a valid state and should be destroyed

Reply:
https://docs.rs/miniscript/7.0.0/miniscript/psbt/trait.PsbtExt.html

note that the approach we took with PsbtExt in rust-miniscript is to provide both through a single trait, depending on waht you want.